### PR TITLE
Portable text

### PR DESCRIPTION
**Pull request by pokornyd**
### Motivation

This PR introduces portable text support to the rich text parser. Still a work in progress, portable text is currently only supported in RichTextNodeParser. Code base is going to be refactored and functionality extended to RichTextBrowserParser in later iterations. Draft PR created to provide an option to test and review existing functionality.

### Checklist

- [ ] Code follows coding conventions held in this repo
- [ ] Automated tests have been added
- [ ] Tests are passing
- [ ] Docs have been updated (if applicable)
- [ ] Temporary settings (e.g. variables used during development and testing) have been reverted to defaults

### How to test

Instantiate a RichTextNodeParser class, use the parse method on a piece of rich text from Kontent.ai. Try creating obscure edge cases to identify potential issues with the parsing logic. Check existing portable text tests for usage details.

### CODE REVIEW

**Review by JiriLojda — CHANGES_REQUESTED** [review]
Good job with using the FP structure, it looks great. 👍 Refactoring the algorithm to a single recursion through the tree might improve readability and performance. I would also unify the file names as right now there are at least two different naming conventions without an apparent pattern.

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+type TransformLinkFunction = (node: IDomHtmlNode) => [(IPortableTextExternalLink | IPortableTextInternalLink), IPortableTextMark];
+type TransformElementFunction = (node: IDomHtmlNode) => IPortableTextItem[];
+type TransformListItemFunction = (node: IDomHtmlNode, depth: number, listType: ListType) => IPortableTextItem[];
+type TransformTextFunction = (node: IDomTextNode) => IPortableTextSpan;
+
+const blockElements = ['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'];
+const markElements = ['strong', 'em', 'sub', 'sup'];
+const ignoredElements = ['img', 'tbody', 'ol', 'ul'];
+
+export const transform = (parsedTree: IOutputResult): IPortableTextItem[] => {
```
[the remark this answers is not in the document]
Comment (Simply007) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:15]: Could be split out to separate issue since it does not affect public API.

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    const flattened = flatten(parsedTree.children, 0);
+    return composeAndMerge(flattened);
+}
+
+export const mergeSpansAndMarks = (itemsToMerge: IPortableTextItem[]): IPortableTextItem[] => {
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts]: isn't anymore :)

File: test/rich-text-to-json-converter.spec.ts [hunk 3]
```diff
@@ -34,153 +39,165 @@ const dummyRichText: Elements.RichTextElement = {
...
+    })
+    expect(result).toMatchSnapshot();
+  })
 
-    expect(result).toMatchInlineSnapshot(`
-"<p  class=\\"test\\" id=\\"3\\">
-              <div>
-                resolved type: test, value of text element: random text value
-              </div>
+  it("resolves a table", () => {
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply test/rich-text-to-json-converter.spec.ts]: not anymore!

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts]: https://github.com/pokornyd/rich-text-resolver/issues/11

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    ),
+    'a': transformLink,
+    'li': transformListItem,
+    'table': transformTable,
+    'tr': transformTableRow,
+    'td': transformTableCell,
+    'br': transformLineBreak,
+    'figure': transformImage,
+    'object': transformItem,
+    default: (node: IDomHtmlNode) => {
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:311]: I agree that it's preferable to throw an error. Ignoring the tag has the unwanted outcome of subsequent nested spans or blocks being processed anyway, which may result in unexpected issues during the composition process.

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+
+export const mergeSpansAndMarks = (itemsToMerge: IPortableTextItem[]): IPortableTextItem[] => {
+    let marks: string[] = [];
+    let links: string[] = [];
+
+    const mergedItems = itemsToMerge.reduce<IPortableTextItem[]>((mergedItems, item) => {
+        switch (item._type) {
+            case 'internalLink':
+            case 'link':
+                const lastBlockIndex = findLastIndex(mergedItems, item => item._type === 'block');
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:30]: submitted as issue: https://github.com/pokornyd/rich-text-resolver/issues/12

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        return mergedItems;
+    },[])
+
+    return mergedItems;
+}
+
+const composeAndMerge = compose(mergeTablesAndRows, mergeRowsAndCells, mergeCellsAndBlocks, mergeBlocksAndSpans, mergeSpansAndMarks);
+  
+const flatten = (nodes: IDomNode[], depth: number = 0, listType?: ListType): IPortableTextItem[] => {
+    return nodes.reduce((finishedItems: IPortableTextItem[], node: IDomNode) => {
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:137]: agreed, rewritten

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+const ignoredElements = ['img', 'tbody', 'ol', 'ul'];
+
+export const transform = (parsedTree: IOutputResult): IPortableTextItem[] => {
+    const flattened = flatten(parsedTree.children, 0);
+    return composeAndMerge(flattened);
+}
+
+export const mergeSpansAndMarks = (itemsToMerge: IPortableTextItem[]): IPortableTextItem[] => {
+    let marks: string[] = [];
+    let links: string[] = [];
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:23]: I'll address this as a part of https://github.com/pokornyd/rich-text-resolver/issues/11

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+const composeAndMerge = compose(mergeTablesAndRows, mergeRowsAndCells, mergeCellsAndBlocks, mergeBlocksAndSpans, mergeSpansAndMarks);
+  
+const flatten = (nodes: IDomNode[], depth: number = 0, listType?: ListType): IPortableTextItem[] => {
+    return nodes.reduce((finishedItems: IPortableTextItem[], node: IDomNode) => {
+        let transformedNode: IPortableTextItem[] | null = null;
+        let children: IDomNode[] = [];
+        let transformedChildren: IPortableTextItem[] = [];
+        let currentListType = listType;
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:141]: likewise, will be adressed in https://github.com/pokornyd/rich-text-resolver/issues/11 which will ideally render the flatten function obsolete

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+            [tagName]: transformBlock,
+        }),{}
+    ),
+    ...markElements.reduce(
+        (acc, tagName) => ({
+            ...acc,
+            [tagName]: transformTextMark,
+        }), {}
+    ),
+    ...ignoredElements.reduce(
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:297]: rewritten

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+    [createMark(crypto.randomUUID(), node.tagName, 'mark')];
+
+const transformLineBreak: TransformElementFunction = (): IPortableTextSpan[] =>
+    [createSpan(crypto.randomUUID(), [], '\n')];
+
+const transformListItem: TransformListItemFunction = (node: IDomHtmlNode, depth: number, listType: ListType): IPortableTextListBlock[] =>
+    [createListBlock(crypto.randomUUID(), depth, listType!)];
+
+
+const ignoreElement: TransformElementFunction = (node: IDomHtmlNode) => [];
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:282]: oversight, removed :)

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+    ),
+    ...ignoredElements.reduce(
+        (acc, tagName) => ({
+            ...acc,
+            [tagName]: ignoreElement,
+        }), {}
+    ),
+    'a': transformLink,
+    'li': transformListItem,
+    'table': transformTable,
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:305]: It would appear so. Our tables don't really allow any customization apart from formatting the cell contents themselves.

File: src/transformers/portable_text_transformer/PortableTextTransformer.ts [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+const transformBlock: TransformElementFunction = (node: IDomHtmlNode): IPortableTextBlock[] => 
+    [createBlock(crypto.randomUUID(), undefined, node.tagName === 'p' ? 'normal' : node.tagName)];
+
+const transformTextMark: TransformElementFunction = (node: IDomHtmlNode): IPortableTextMark[] =>
+    [createMark(crypto.randomUUID(), node.tagName, 'mark')];
+
+const transformLineBreak: TransformElementFunction = (): IPortableTextSpan[] =>
+    [createSpan(crypto.randomUUID(), [], '\n')];
+
+const transformListItem: TransformListItemFunction = (node: IDomHtmlNode, depth: number, listType: ListType): IPortableTextListBlock[] =>
```
[the remark this answers is not in the document]
Comment (pokornyd) [reply src/transformers/portable_text_transformer/PortableTextTransformer.ts:279]: thanks, replaced the unused argument :)
